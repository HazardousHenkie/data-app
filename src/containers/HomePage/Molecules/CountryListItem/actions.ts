import { action } from 'typesafe-actions'

import ActionTypes from './constants'

import CountryState from './types'

const setSelectedCountry = (country: CountryState['country']) =>
    action(ActionTypes.SET_SELECTED_COUNTRY, country)

export default setSelectedCountry
