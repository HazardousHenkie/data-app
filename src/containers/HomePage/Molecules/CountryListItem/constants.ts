import CountryState from './types'

enum ActionTypes {
    SET_SELECTED_COUNTRY = 'components/CountriesList/SET_SELECTED_COUNTRY'
}

export const CountryItem: CountryState = {
    country: {
        alpha2Code: '',
        name: '',
        nativeName: '',
        capital: '',
        region: '',
        subregion: '',
        flag: '',
        currency: '',
        population: 0,
        latlng: [0, 0],
        currencies: [{ string: '' }],
        languages: [{ string: '' }],
        translations: { string: '' }
    }
}

export default ActionTypes
