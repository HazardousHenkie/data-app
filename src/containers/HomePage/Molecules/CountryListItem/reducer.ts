import { ContainerCountryState, ContainerCountryActions } from './types'

import ActionTypes from './constants'

export const initialCountryState: ContainerCountryState = {
    country: {}
}

function countriesListReducer(
    state: ContainerCountryState = initialCountryState,
    action: ContainerCountryActions
): ContainerCountryState {
    switch (action.type) {
        case ActionTypes.SET_SELECTED_COUNTRY:
            return {
                country: action.payload
            }
        default:
            return state
    }
}

export default countriesListReducer
