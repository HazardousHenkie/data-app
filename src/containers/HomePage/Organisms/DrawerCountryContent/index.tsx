import React from 'react'

import { createSelector } from 'reselect'
import { useSelector } from 'react-redux'
import { useTranslation } from 'react-i18next'

import makeSelectCountry from 'containers/HomePage/Molecules/CountryListItem/selectors'

import Grid from '@material-ui/core/Grid'

import FavoriteCountryButton from 'components/Organisms/FavoriteCountryButton'
import Header, { FavoriteCountryButtonWrapper } from './styledComponents'
import CountryAdvisory from '../../Molecules/CountryAdvisory'
import CountryInformation from '../../Molecules/CountryInformation'

const stateSelector = createSelector(makeSelectCountry(), country => ({
    country
}))

const DrawerCountryContent: React.FC = () => {
    const { country } = useSelector(stateSelector)
    const { i18n } = useTranslation()

    return (
        <>
            <Grid container spacing={3}>
                <Grid item xs={12}>
                    <Header>
                        {i18n.language === 'en'
                            ? country.name
                            : country.translations[i18n.language]}
                    </Header>

                    <FavoriteCountryButtonWrapper>
                        <FavoriteCountryButton clickedCountry={country} />
                    </FavoriteCountryButtonWrapper>
                </Grid>

                <Grid item xs={6}>
                    <CountryInformation />
                </Grid>

                <Grid item xs={6}>
                    <CountryAdvisory />
                </Grid>
            </Grid>
        </>
    )
}

export default DrawerCountryContent
