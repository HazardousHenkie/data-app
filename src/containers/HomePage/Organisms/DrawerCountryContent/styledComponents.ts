import styled from 'styled-components'

const HeaderStyled = styled.h1`
    font-size: 2rem;
    text-align: center;
    margin: 0 0 10px;
`

export const FavoriteCountryButtonWrapper = styled.div`
    position: absolute;
    right: 20px;
    top: 20px;
`

export default HeaderStyled
