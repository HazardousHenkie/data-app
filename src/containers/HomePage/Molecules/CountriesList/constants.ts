import { CountryItem } from 'containers/HomePage/Molecules/CountryListItem/constants'
import CountriesListState from './types'

enum ActionTypes {
    GET_COUNTRIES_DATA = 'components/CountriesList/GET_COUNTRIES_DATA',
    GET_COUNTRIES_DATA_SUCCESS = 'components/CountriesList/GET_COUNTRIES_DATA_SUCCESS',
    GET_COUNTRIES_DATA_ERROR = 'components/CountriesList/GET_COUNTRIES_DATA_ERROR'
}

export const initialCountriesHeaderState: CountriesListState = {
    error: false,
    loading: false,
    data: [CountryItem.country]
}

export default ActionTypes
