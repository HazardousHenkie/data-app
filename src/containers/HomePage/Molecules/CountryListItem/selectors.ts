import { createSelector, Selector } from 'reselect'

import { ApplicationRootState } from 'types'
import { CountryInterface } from './types'
import { CountryItem } from './constants'

export const initialCountryState = CountryItem

export const selectCountry = (state: ApplicationRootState) => {
    return state.country || initialCountryState
}

const makeSelectCountry = () =>
    createSelector(selectCountry, subState => subState.country) as Selector<
        unknown,
        CountryInterface
    >

export default makeSelectCountry
