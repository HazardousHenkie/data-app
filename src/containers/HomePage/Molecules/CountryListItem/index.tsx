import React from 'react'

import { Reducer } from 'redux'

import { useInjectReducer } from 'utils/redux-injectors'

import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'
import ListItemIcon from '@material-ui/core/ListItemIcon'

import { useTranslation } from 'react-i18next'

import { useDispatch, useSelector } from 'react-redux'
import { createSelector } from 'reselect'
import { makeSelectData } from 'containers/HomePage/Molecules/CountriesList/selectors'
import { CountryInterface } from './types'
import setSelectedCountry from './actions'

import reducer from './reducer'

interface CountriesListItemProps {
    setOpen: React.Dispatch<React.SetStateAction<boolean>>
    listCountry: CountryInterface
}

const key = 'country'

const stateSelector = createSelector(makeSelectData(), countries => ({
    countries
}))

const CountriesListItem: React.FC<CountriesListItemProps> = ({
    setOpen,
    listCountry
}) => {
    const dispatch = useDispatch()
    const { i18n } = useTranslation()
    const { countries } = useSelector(stateSelector)

    useInjectReducer({ key, reducer: reducer as Reducer })

    const onClickCountry = (selectedName: string) => {
        const selectedCountry = countries.find(
            country => country.name === selectedName
        )

        if (selectedCountry) {
            setOpen(false)
            dispatch(setSelectedCountry(selectedCountry))
        }
    }

    return (
        <ListItem
            data-testid="CountriesListItem"
            button
            onClick={() => onClickCountry(listCountry.name)}
        >
            <ListItemIcon data-testid="CountriesListItemIcon">
                <img
                    width="30"
                    height="20"
                    src={listCountry.flag}
                    alt={listCountry.name}
                />
            </ListItemIcon>
            <ListItemText
                data-testid="CountriesListItemText"
                primary={
                    i18n.language === 'en'
                        ? listCountry.name
                        : listCountry.translations[i18n.language]
                }
            />
        </ListItem>
    )
}

export default CountriesListItem
