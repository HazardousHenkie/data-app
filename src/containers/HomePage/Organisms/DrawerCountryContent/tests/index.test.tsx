import React from 'react'
import { render } from 'utils/test-utils'

import configureStore from 'store/configureStore'
import history from 'utils/history'

import { CountryItem } from 'containers/HomePage/Molecules/CountryListItem/constants'
import setSelectedCountry from 'containers/HomePage/Molecules/CountryListItem/actions'

import DrawerCountryContent from '../index'

describe('<FavoriteCountryButton />', () => {
    it('should render like snapshot', () => {
        const component = render(<DrawerCountryContent />)

        expect(component).toMatchSnapshot()
    })

    it('should render Header with countryName', () => {
        const mockStore = configureStore({}, history)

        const { getByText } = render(<DrawerCountryContent />, {
            store: mockStore
        })

        mockStore.dispatch(
            setSelectedCountry({ ...CountryItem.country, name: 'japan' })
        )

        const header = getByText('japan')

        expect(header).toBeInTheDocument()
        expect(header.tagName).toBe('H1')
    })

    it('should render FavoriteCountryButtonWrapper', () => {
        const { getByTestId } = render(<DrawerCountryContent />, {
            initialState: { authenticationData: { loggedIn: true } }
        })

        const FavoriteCountryButtonWrapper = getByTestId(
            'FavoriteCountryButtonWrapper'
        )

        expect(FavoriteCountryButtonWrapper).toBeInTheDocument()
    })

    it('should render heartButton', () => {
        const { getByTestId } = render(<DrawerCountryContent />, {
            initialState: {
                authenticationData: { loggedIn: true }
            }
        })

        const heartButton = getByTestId('heartButton')

        expect(heartButton).toBeInTheDocument()
    })

    it('should render CountryAdvisory', () => {
        const { getByTestId } = render(<DrawerCountryContent />)

        const CountryAdvisory = getByTestId('CountryAdvisory')

        expect(CountryAdvisory).toBeInTheDocument()
    })

    it('should render CountryInformationCard', () => {
        const { getByTestId } = render(<DrawerCountryContent />)

        const CountryInformationCard = getByTestId('CountryInformationCard')

        expect(CountryInformationCard).toBeInTheDocument()
    })
})
