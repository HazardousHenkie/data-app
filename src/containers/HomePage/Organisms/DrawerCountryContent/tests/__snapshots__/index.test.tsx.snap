// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavoriteCountryButton /> should render like snapshot 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
  >
    <h1
      className="sc-AxhUy idbwSw"
    />
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
  >
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        className="MuiCardHeader-root"
      >
        <div
          className="MuiCardHeader-avatar"
        >
          <div
            aria-label="homePage:countryAdvisor.countryInformation"
            className="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
              />
            </svg>
          </div>
        </div>
        <div
          className="MuiCardHeader-content"
        >
          <span
            className="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
          >
            homePage:countryAdvisor.advisoryTitle
                undefined
          </span>
        </div>
      </div>
      <div
        className="MuiCardContent-root"
      >
        
        
        
        
        0
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          <strong>
            homePage:countryInformation.languages
            :
          </strong>
            undefined(undefined)
        </p>
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          <strong>
            homePage:countryInformation.currencies
            :
          </strong>
            undefined(undefined)
        </p>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
  >
    <div
      className="MuiPaper-root MuiCard-root sc-AxheI YnnDL MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        className="MuiCardHeader-root"
      >
        <div
          className="MuiCardHeader-avatar"
        >
          <div
            aria-label="homePage:countryAdvisor.avatar"
            className="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
              />
            </svg>
          </div>
        </div>
        <div
          className="MuiCardHeader-content"
        >
          <span
            className="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
          >
            homePage:countryAdvisor.advisoryTitle
                undefined
          </span>
        </div>
      </div>
      <div
        className="MuiCardContent-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          homePage:countryAdvisor.notFound
        </p>
      </div>
    </div>
  </div>
</div>
`;
